File: clients/go/model/building_properties.go
package model

import "fmt"
import "io"
import . "spb_ai_champ/stream"

// TODO - Document
type BuildingProperties struct {
    // TODO - Document
    BuildResources map[Resource]int32
    // TODO - Document
    MaxHealth int32
    // TODO - Document
    MaxWorkers int32
    // TODO - Document
    WorkResources map[Resource]int32
    // TODO - Document
    ProduceWorker bool
    // TODO - Document
    ProduceResource *Resource
    // TODO - Document
    ProduceAmount int32
    // TODO - Document
    ProduceScore int32
    // TODO - Document
    Harvest bool
    // TODO - Document
    WorkAmount int32
}

func NewBuildingProperties(buildResources map[Resource]int32, maxHealth int32, maxWorkers int32, workResources map[Resource]int32, produceWorker bool, produceResource *Resource, produceAmount int32, produceScore int32, harvest bool, workAmount int32) BuildingProperties {
    return BuildingProperties {
        BuildResources: buildResources,
        MaxHealth: maxHealth,
        MaxWorkers: maxWorkers,
        WorkResources: workResources,
        ProduceWorker: produceWorker,
        ProduceResource: produceResource,
        ProduceAmount: produceAmount,
        ProduceScore: produceScore,
        Harvest: harvest,
        WorkAmount: workAmount,
    }
}

// Read BuildingProperties from reader
func ReadBuildingProperties(reader io.Reader) BuildingProperties {
    var buildResources map[Resource]int32
    buildResourcesSize := ReadInt32(reader)
    buildResources = make(map[Resource]int32)
    for buildResourcesIndex := int32(0); buildResourcesIndex < buildResourcesSize; buildResourcesIndex++ {
        var buildResourcesKey Resource
        buildResourcesKey = ReadResource(reader)
        var buildResourcesValue int32
        buildResourcesValue = ReadInt32(reader)
        buildResources[buildResourcesKey] = buildResourcesValue
    }
    var maxHealth int32
    maxHealth = ReadInt32(reader)
    var maxWorkers int32
    maxWorkers = ReadInt32(reader)
    var workResources map[Resource]int32
    workResourcesSize := ReadInt32(reader)
    workResources = make(map[Resource]int32)
    for workResourcesIndex := int32(0); workResourcesIndex < workResourcesSize; workResourcesIndex++ {
        var workResourcesKey Resource
        workResourcesKey = ReadResource(reader)
        var workResourcesValue int32
        workResourcesValue = ReadInt32(reader)
        workResources[workResourcesKey] = workResourcesValue
    }
    var produceWorker bool
    produceWorker = ReadBool(reader)
    var produceResource *Resource
    if ReadBool(reader) {
        var produceResourceValue Resource
        produceResourceValue = ReadResource(reader)
        produceResource = &produceResourceValue
    } else {
        produceResource = nil
    }
    var produceAmount int32
    produceAmount = ReadInt32(reader)
    var produceScore int32
    produceScore = ReadInt32(reader)
    var harvest bool
    harvest = ReadBool(reader)
    var workAmount int32
    workAmount = ReadInt32(reader)
    return BuildingProperties {
        BuildResources: buildResources,
        MaxHealth: maxHealth,
        MaxWorkers: maxWorkers,
        WorkResources: workResources,
        ProduceWorker: produceWorker,
        ProduceResource: produceResource,
        ProduceAmount: produceAmount,
        ProduceScore: produceScore,
        Harvest: harvest,
        WorkAmount: workAmount,
    }
}

// Write BuildingProperties to writer
func (buildingProperties BuildingProperties) Write(writer io.Writer) {
    buildResources := buildingProperties.BuildResources
    WriteInt32(writer, int32(len(buildResources)))
    for buildResourcesKey, buildResourcesValue := range buildResources {
        WriteInt32(writer, int32(buildResourcesKey))
        WriteInt32(writer, buildResourcesValue)
    }
    maxHealth := buildingProperties.MaxHealth
    WriteInt32(writer, maxHealth)
    maxWorkers := buildingProperties.MaxWorkers
    WriteInt32(writer, maxWorkers)
    workResources := buildingProperties.WorkResources
    WriteInt32(writer, int32(len(workResources)))
    for workResourcesKey, workResourcesValue := range workResources {
        WriteInt32(writer, int32(workResourcesKey))
        WriteInt32(writer, workResourcesValue)
    }
    produceWorker := buildingProperties.ProduceWorker
    WriteBool(writer, produceWorker)
    produceResource := buildingProperties.ProduceResource
    if produceResource == nil {
        WriteBool(writer, false)
    } else {
        WriteBool(writer, true)
        produceResourceValue := *produceResource
        WriteInt32(writer, int32(produceResourceValue))
    }
    produceAmount := buildingProperties.ProduceAmount
    WriteInt32(writer, produceAmount)
    produceScore := buildingProperties.ProduceScore
    WriteInt32(writer, produceScore)
    harvest := buildingProperties.Harvest
    WriteBool(writer, harvest)
    workAmount := buildingProperties.WorkAmount
    WriteInt32(writer, workAmount)
}

// Get string representation of BuildingProperties
func (buildingProperties BuildingProperties) String() string {
    stringResult := "{ "
    stringResult += "BuildResources: "
    buildResources := buildingProperties.BuildResources
    stringResult += "map[ "
    buildResourcesIndex := 0
    for buildResourcesKey, buildResourcesValue := range buildResources {
        if buildResourcesIndex != 0 {
            stringResult += ", "
        }
        stringResult += ResourceToString(buildResourcesKey)
        stringResult += ": "
        stringResult += fmt.Sprint(buildResourcesValue)
        buildResourcesIndex++
    }
    stringResult += " ]"
    stringResult += ", "
    stringResult += "MaxHealth: "
    maxHealth := buildingProperties.MaxHealth
    stringResult += fmt.Sprint(maxHealth)
    stringResult += ", "
    stringResult += "MaxWorkers: "
    maxWorkers := buildingProperties.MaxWorkers
    stringResult += fmt.Sprint(maxWorkers)
    stringResult += ", "
    stringResult += "WorkResources: "
    workResources := buildingProperties.WorkResources
    stringResult += "map[ "
    workResourcesIndex := 0
    for workResourcesKey, workResourcesValue := range workResources {
        if workResourcesIndex != 0 {
            stringResult += ", "
        }
        stringResult += ResourceToString(workResourcesKey)
        stringResult += ": "
        stringResult += fmt.Sprint(workResourcesValue)
        workResourcesIndex++
    }
    stringResult += " ]"
    stringResult += ", "
    stringResult += "ProduceWorker: "
    produceWorker := buildingProperties.ProduceWorker
    stringResult += fmt.Sprint(produceWorker)
    stringResult += ", "
    stringResult += "ProduceResource: "
    produceResource := buildingProperties.ProduceResource
    if produceResource == nil {
        stringResult += "nil"
    } else {
        produceResourceValue := *produceResource
        stringResult += ResourceToString(produceResourceValue)
    }
    stringResult += ", "
    stringResult += "ProduceAmount: "
    produceAmount := buildingProperties.ProduceAmount
    stringResult += fmt.Sprint(produceAmount)
    stringResult += ", "
    stringResult += "ProduceScore: "
    produceScore := buildingProperties.ProduceScore
    stringResult += fmt.Sprint(produceScore)
    stringResult += ", "
    stringResult += "Harvest: "
    harvest := buildingProperties.Harvest
    stringResult += fmt.Sprint(harvest)
    stringResult += ", "
    stringResult += "WorkAmount: "
    workAmount := buildingProperties.WorkAmount
    stringResult += fmt.Sprint(workAmount)
    stringResult += " }"
    return stringResult
}

File: clients/csharp/Model/BuildingProperties.cs
namespace SpbAiChamp.Model
{
    /// <summary>
    /// TODO - Document
    /// </summary>
    public struct BuildingProperties
    {
        /// <summary>
        /// TODO - Document
        /// </summary>
        public System.Collections.Generic.IDictionary<Model.Resource, int> BuildResources { get; set; }
        /// <summary>
        /// TODO - Document
        /// </summary>
        public int MaxHealth { get; set; }
        /// <summary>
        /// TODO - Document
        /// </summary>
        public int MaxWorkers { get; set; }
        /// <summary>
        /// TODO - Document
        /// </summary>
        public System.Collections.Generic.IDictionary<Model.Resource, int> WorkResources { get; set; }
        /// <summary>
        /// TODO - Document
        /// </summary>
        public bool ProduceWorker { get; set; }
        /// <summary>
        /// TODO - Document
        /// </summary>
        public Model.Resource? ProduceResource { get; set; }
        /// <summary>
        /// TODO - Document
        /// </summary>
        public int ProduceAmount { get; set; }
        /// <summary>
        /// TODO - Document
        /// </summary>
        public int ProduceScore { get; set; }
        /// <summary>
        /// TODO - Document
        /// </summary>
        public bool Harvest { get; set; }
        /// <summary>
        /// TODO - Document
        /// </summary>
        public int WorkAmount { get; set; }
    
        public BuildingProperties(System.Collections.Generic.IDictionary<Model.Resource, int> buildResources, int maxHealth, int maxWorkers, System.Collections.Generic.IDictionary<Model.Resource, int> workResources, bool produceWorker, Model.Resource? produceResource, int produceAmount, int produceScore, bool harvest, int workAmount)
        {
            this.BuildResources = buildResources;
            this.MaxHealth = maxHealth;
            this.MaxWorkers = maxWorkers;
            this.WorkResources = workResources;
            this.ProduceWorker = produceWorker;
            this.ProduceResource = produceResource;
            this.ProduceAmount = produceAmount;
            this.ProduceScore = produceScore;
            this.Harvest = harvest;
            this.WorkAmount = workAmount;
        }
    
        /// <summary> Read BuildingProperties from reader </summary>
        public static BuildingProperties ReadFrom(System.IO.BinaryReader reader)
        {
            var result = new BuildingProperties();
            int buildResourcesSize = reader.ReadInt32();
            result.BuildResources = new System.Collections.Generic.Dictionary<Model.Resource, int>(buildResourcesSize);
            for (int buildResourcesIndex = 0; buildResourcesIndex < buildResourcesSize; buildResourcesIndex++)
            {
                Model.Resource buildResourcesKey;
                int buildResourcesValue;
                buildResourcesKey = ResourceHelper.ReadFrom(reader);
                buildResourcesValue = reader.ReadInt32();
                result.BuildResources.Add(buildResourcesKey, buildResourcesValue);
            }
            result.MaxHealth = reader.ReadInt32();
            result.MaxWorkers = reader.ReadInt32();
            int workResourcesSize = reader.ReadInt32();
            result.WorkResources = new System.Collections.Generic.Dictionary<Model.Resource, int>(workResourcesSize);
            for (int workResourcesIndex = 0; workResourcesIndex < workResourcesSize; workResourcesIndex++)
            {
                Model.Resource workResourcesKey;
                int workResourcesValue;
                workResourcesKey = ResourceHelper.ReadFrom(reader);
                workResourcesValue = reader.ReadInt32();
                result.WorkResources.Add(workResourcesKey, workResourcesValue);
            }
            result.ProduceWorker = reader.ReadBoolean();
            if (reader.ReadBoolean())
            {
                result.ProduceResource = ResourceHelper.ReadFrom(reader);
            } else
            {
                result.ProduceResource = null;
            }
            result.ProduceAmount = reader.ReadInt32();
            result.ProduceScore = reader.ReadInt32();
            result.Harvest = reader.ReadBoolean();
            result.WorkAmount = reader.ReadInt32();
            return result;
        }
    
        /// <summary> Write BuildingProperties to writer </summary>
        public void WriteTo(System.IO.BinaryWriter writer)
        {
            writer.Write(BuildResources.Count);
            foreach (var buildResourcesEntry in BuildResources)
            {
                var buildResourcesKey = buildResourcesEntry.Key;
                var buildResourcesValue = buildResourcesEntry.Value;
                writer.Write((int) (buildResourcesKey));
                writer.Write(buildResourcesValue);
            }
            writer.Write(MaxHealth);
            writer.Write(MaxWorkers);
            writer.Write(WorkResources.Count);
            foreach (var workResourcesEntry in WorkResources)
            {
                var workResourcesKey = workResourcesEntry.Key;
                var workResourcesValue = workResourcesEntry.Value;
                writer.Write((int) (workResourcesKey));
                writer.Write(workResourcesValue);
            }
            writer.Write(ProduceWorker);
            if (!ProduceResource.HasValue)
            {
                writer.Write(false);
            } else
            {
                writer.Write(true);
                writer.Write((int) (ProduceResource.Value));
            }
            writer.Write(ProduceAmount);
            writer.Write(ProduceScore);
            writer.Write(Harvest);
            writer.Write(WorkAmount);
        }
    
        /// <summary> Get string representation of BuildingProperties </summary>
        public override string ToString() {
            string stringResult = "BuildingProperties { ";
            stringResult += "BuildResources: ";
            stringResult += "{ ";
            int buildResourcesIndex = 0;
            foreach (var buildResourcesEntry in BuildResources)
            {
                if (buildResourcesIndex != 0) {
                    stringResult += ", ";
                }
                var buildResourcesKey = buildResourcesEntry.Key;
                stringResult += buildResourcesKey.ToString();
                stringResult += ": ";
                var buildResourcesValue = buildResourcesEntry.Value;
                stringResult += buildResourcesValue.ToString();
                buildResourcesIndex++;
            }
            stringResult += " }";
            stringResult += ", ";
            stringResult += "MaxHealth: ";
            stringResult += MaxHealth.ToString();
            stringResult += ", ";
            stringResult += "MaxWorkers: ";
            stringResult += MaxWorkers.ToString();
            stringResult += ", ";
            stringResult += "WorkResources: ";
            stringResult += "{ ";
            int workResourcesIndex = 0;
            foreach (var workResourcesEntry in WorkResources)
            {
                if (workResourcesIndex != 0) {
                    stringResult += ", ";
                }
                var workResourcesKey = workResourcesEntry.Key;
                stringResult += workResourcesKey.ToString();
                stringResult += ": ";
                var workResourcesValue = workResourcesEntry.Value;
                stringResult += workResourcesValue.ToString();
                workResourcesIndex++;
            }
            stringResult += " }";
            stringResult += ", ";
            stringResult += "ProduceWorker: ";
            stringResult += ProduceWorker.ToString();
            stringResult += ", ";
            stringResult += "ProduceResource: ";
            if (!ProduceResource.HasValue)
            {
                stringResult += "null";
            } else
            {
                stringResult += ProduceResource.Value.ToString();
            }
            stringResult += ", ";
            stringResult += "ProduceAmount: ";
            stringResult += ProduceAmount.ToString();
            stringResult += ", ";
            stringResult += "ProduceScore: ";
            stringResult += ProduceScore.ToString();
            stringResult += ", ";
            stringResult += "Harvest: ";
            stringResult += Harvest.ToString();
            stringResult += ", ";
            stringResult += "WorkAmount: ";
            stringResult += WorkAmount.ToString();
            stringResult += " }";
            return stringResult;
        }
    }
}